### PR TITLE
MLv2: Remove `metabase.lib.dev` and "unresolved" functions 

### LINKED ISSUE #30041: [MLv2] external op format as output should be returned by a separate function
@metamben and I discussed this today in our 1:1

We (probably) will still need external ops to edit existing filter clauses:

image_480

but since editing is the exceptional case rather than the norm, something like `filters(query)` should probably just return opaque MBQL filter clauses, so you can call `display-name` or stuff like that on them if all you need to do is render them, e.g.:

image_360

If you need to **edit** them, you can pass in an opaque clause to some sort of `externalOp(clause)` function to get the external op format that is easily editable.

### LINKED ISSUE #31741: [MLv2] Several aggregation schemas incorrectly only allow numeric args
`:max` and `:min` have schemas like

```clj
(mbql-clause/define-tuple-mbql-clause :max
  [:schema [:ref ::expression/number]])
```

However, this is incorrect, since you should be able to do `max(<date>)` -- we should use `::expression/orderable` rather than `::expression/number`.

Also applies to 

Similarly, `:count` and and `:cum-count`, require their args to be numbers, but these simply map to the `count(<expr>)` functions in SQL or equivalent, that count the number of non-null rows.

We should go thru these expression schemas and try to make sure we're using the correct types for things.

### PR DESCRIPTION
**Pull request by camsaul**
Remove the `metabase.lib.dev` stuff and thus the need to support unresolved functions for things. As a result we don't need to pass in `query` or `stage-number` to a lot of places where it was previously required, for example `order-by-clause` or `count` or `=` or stuff like that. In turn, this makes things easier to play around with in the REPL, because something like `(lib/= <field>)` returns a `[:= ...]` clause now rather than a function that needs to be resolved. 

I added new `expression-ref` and `aggregation-ref` functions that replace the places we used `lib.dev/ref-lookup`. These take query and (optionally) stage number and return a resolved reference, rather than a function. We now need to use `as->` and pass in the query when we use them. 

Regarding other `lib.dev` stuff: we can use mock metadata or the like directly for Field/column metadata and Table metadata, instead of `lib.dev/field` and `lib.dev/table`, respectively. At some point in the future when we start working on porting the QP we can add new helper functions or macros for app DB use (e.g. "get Column metadata for Field X for `(mt/db)`;", "get Table metadata for Table X") so it should not make working on stuff there any more annoying.

This also lets us do better type schema checking, because now most things are "takes an X and returns a Y" rather than "takes an X **or** an unchecked function and returns a Y **or** an unchecked function". This caught the bug #31741, which I've fixed in this PR.

Fixes #31741

As part of this PR I was able to drastically simplify the functions generated by `defop` for things like `lib/count` and `lib/=`, and I went ahead and changed them to return raw MBQL clauses rather than the external op format since that was something we wanted to do anyway in #30041 

Fixes #30041

### CODE REVIEW

File: src/metabase/lib/expression.cljc [hunk 1]
```diff
@@ -241,6 +248,16 @@
 (lib.common/defop upper [s])
 (lib.common/defop lower [s])
 
+(mu/defn ^:private expression-metadata :- lib.metadata/ColumnMetadata
```
Comment (camsaul) [inline src/metabase/lib/expression.cljc:251]: Moved from the bottom of the file but otherwise unchanged

File: src/metabase/lib/join.cljc [hunk 2]
```diff
@@ -187,77 +184,55 @@
...
+  {:arglists '([joinable])}
+  lib.dispatch/dispatch-value
   :hierarchy lib.hierarchy/hierarchy)
 
 ;; TODO -- should the default implementation call [[metabase.lib.query/query]]? That way if we implement a method to
 ;; create an MBQL query from a `Table`, then we'd also get [[join]] support for free?
 
 (defmethod join-clause-method :mbql/join
-  [_query _stage-number a-join-clause]
+  [a-join-clause]
```
Comment (camsaul) [inline src/metabase/lib/join.cljc:195]: Nice 💪

**Review by bshepherdson — APPROVED** [review]
I love it! I think this simplifies a complicated part of the machinery, and nearly all the tests are 1-1, just changing how things are "spelled".

**Review by metamben — APPROVED** [review]
👍

**Pull request merged**